Update wxAuiGenericTabArt and wxAuiSimpleTabArt to properly handle sc…

pkulchenko opened the pull request:
…aled content.

This fixes both Generic and Simple tab arts. Here are the screenshots for both tab arts with the scale factor set at 3 on win10:

![wxwidgets-aui-scale-3-generic-fix](https://user-images.githubusercontent.com/77071/66626377-dc092580-ebac-11e9-9ef4-d67c187a30e6.png)

![wxwidgets-aui-scale-3-simple-fix](https://user-images.githubusercontent.com/77071/66626382-e3303380-ebac-11e9-96ea-1ec0433f9386.png)

Everything looks good so far and I've also tested with scalefactor=2 and confirmed that the changes look correct on hidpi screen on macos.

Couple of questions. Not sure what the best way to check if the scaling needs to be applied, but I added `wnd->FromDIP(1) > 1.0 && (bmp.GetWidth() == 16 || bmp.GetHeight() == 16)` to trigger the scaling. I don't like `==16` part, so if there is a better way to confirm that this is the original bmps, please advise/replace. This can probably be removed completely, but I wanted to protect against those cases where a different (appropriately scaled) image is passed to those routines (in the future).

I also replaced `GetScaledWidth/Height` with `GetWidth/Height` versions. I think it was a mistake, but it wasn't visible when scaling is 1. All the calculations are in screen pixels with fromDIP applied, so it probably doesn't make sense to combine `GetScaled` values with `fromDIP` results. I confirmed that everything looks correct and proportionate on the screen, but would welcome feedback to confirm.

pkulchenko committed:
Update wxAuiGenericTabArt and wxAuiSimpleTabArt to properly handle scaled content.

vadz commented on src/aui/tabart.cpp:
```diff
@@ -221,7 +221,7 @@ void wxAuiGenericTabArt::SetSizingInfo(const wxSize& tab_ctrl_size,
     m_fixedTabWidth = wxWindow::FromDIP(100, NULL);
 
-    int tot_width = (int)tab_ctrl_size.x - GetIndentSize() - 4;
+    int tot_width = (int)tab_ctrl_size.x - GetIndentSize() - wxWindow::FromDIP(4, NULL);
```
This is not a remark for this PR per se, but it's wrong to use `NULL` instead of the actual window here, as this won't work for the windows shown on a secondary display using a different DPI.

We really need to find a way to associate a `wxWindow` with the tab art object, i.e. either pass it as argument to `SetSizingInfo()` or add a separate method to just set the window.

pkulchenko replied:
I tend to agree, but this was done according to the earlier changes by @MaartenBent and there are multiple places where this would need to be updated (with `wnd` dragged through all the calls). I think it's sufficient for this PR to handle it this way and we can deal with these `FromDIP` calls separately.

vadz commented on src/aui/tabart.cpp:
```diff
@@ -1192,11 +1204,12 @@ wxSize wxAuiSimpleTabArt::GetTabSize(wxDC& dc,
 
     if (close_button_state != wxAUI_BUTTON_STATE_HIDDEN)
-        tab_width += m_activeCloseBmp.GetScaledWidth();
+        // increase by button size plus the padding
```
Trivial, but could you please add `{` and `}` around the (now) multiline `if` statement body?

pkulchenko replied:
Yes, I added that.

vadz replied:
This doesn't seem to have been done?

vadz commented on src/aui/tabart.cpp:
```diff
@@ -546,14 +540,49 @@ void wxAuiGenericTabArt::DrawTab(wxDC& dc,
+            wxImage img = bmp.ConvertToImage();
+            img.Rescale(wnd->FromDIP(16), wnd->FromDIP(16), wxIMAGE_QUALITY_BOX_AVERAGE);
+            bmp = img;
+        }
```
This should be extracted in a helper function to avoid repeating it at least thrice.

pkulchenko replied:
Yes, extracted into a separate function and added checks. See the updated PR.

vadz reviewed (dismissed):
Thanks for the PR, but unfortunately I think there are a couple of issues with it that are probably bad enough to prevent applying it for now, even if this change clearly improves things.

First one is that I think we do need to use `GetScaledXXX()` in some places, as this change probably breaks things for Mac/GTK3 where the window and DC coordinates are in logical pixels, while with this PR methods such as `wxDC::DrawRectangle()` would be passed `GetXXX()` that return size in physical pixels.

Second one is that we should keep the scale factor by which our bitmaps have been rescaled instead of checking whether their size is 16x16. This will result in more readable code immediately, i.e. the strange `wnd->FromDIP(1) > 1` test would be replaced with `wnd->GetContentScaleFactor() != m_bitmapScaleFactor` (with `m_bitmapScaleFactor = 1.0` initially), and will also make it more straightforward to support per-monitor DPI later, as we would just need to refresh these elements to redraw them with the bitmaps using correct scale.

@MaartenBent please correct me if I'm wrong on either of these points. TIA!

MaartenBent commented:
I haven't looked into bitmaps, but using `wnd->FromDIP(1) > 1.0` will not work for all scale factors, e.g. when using 150%. `wnd->FromDIP(1)` will still return 1, but the images should be scaled to 24px. This definitely has to be replaced with `GetContentScaleFactor()` as this does return a fractional scale factor.

Maybe you can scale the bitmaps to the correct size in the constructor, so you don't have to check the current size everytime?

Is `wxAuiMSWTabArt` already correct or do you not use it?

MaartenBent commented:
After thinking some more, and considering per-monitor DPI, I think it can't be done in the constructor. For example in a use-case where one TabArt is used for multiple AUI books on different windows with different DPI. So it has to be scaled in the draw functions where the wxWindow is known.
Maybe save the bitmaps in a map, with key=dpi and value=bitmap? So you don't have to scale every time the DPI changes. But that becomes more a memory vs cpu discussion.

pkulchenko committed:
Updated to fix formatting style for multi-line `if`.

pkulchenko committed:
Updated bitmap scaling in wxAui{Generic,Simple}TabArt for better results.

pkulchenko reviewed (commented):
@MaartenBent, I agree that it can't be done in the constructor. I incorporated all the suggestions other than removing ::FromDIP calls, as there are many of them (outside of the PR) that need to be dealt with for consistency. I suggest we handle it when the DPIperMonitor support is added.

pkulchenko commented:
@MaartenBent, to address the other two comments:

> I haven't looked into bitmaps, but using wnd->FromDIP(1) > 1.0 will not work for all scale factors, e.g. when using 125%. wnd->FromDIP(1) will still return 1, but the images should be scaled to 20px. 

This was sort of intentional. I tried rescaling for different scale factors and 125% and 175% looked ugly and %150 was so so. I ended up using scale factor 2 for `1.75 <= scales < 2.75`. The results are acceptable for 125%-350% I tested with and are much better than the current ones.

> So it has to be scaled in the draw functions where the wxWindow is known.
> Maybe save the bitmaps in a map, with key=dpi and value=bitmap? So you don't have to scale every time the DPI changes. But that becomes more a memory vs cpu discussion.

I agree; this is doable, but the current logic works well and if someone wants to extend it with the caching of the changes, that's fine by me. The content of the PR should handle moving between the screens, when it's implemented (other than ::FromDIP calls, but most of those were present before the PR).

pkulchenko committed:
Disabled scaling for wxAui*TabArt on Mac/GTK3.

pkulchenko commented:
> Is wxAuiMSWTabArt already correct or do you not use it?

Yes, looks correct to me, so I didn't apply any changes to it.

I pushed a change to disable scaling on MacOS and GTK3.

vadz commented on src/aui/dockart.cpp:
```diff
@@ -88,6 +89,22 @@ wxBitmap wxAuiBitmapFromBits(const unsigned char bits[], int w, int h,
+#if wxUSE_IMAGE && !defined(__WXGTK3__) && !defined(__WXMAC__)
+    // scale to a close round number to improve quality
+    scale = floor(scale + 0.25);
+    if (scale > 1.0 && !(bmp.GetScaleFactor() > 1.0))
```
I'm still uncomfortable with checking bitmap scale factor as the indication of whether it was scaled or not. We should store the scale factor the bitmap is good for (`1.0` initially) instead.

pkulchenko replied:
I'm not quite sure how this needs to be done (or how this scale factor would be modified). Are you proposing adding a new API call? If you, can you take a stab at this and I'll test the changes?

vadz reviewed (commented):
Have you tested this under the other platforms? I expect macOS will scale the bitmaps on its own, but I'm not sure about GTK.

Also, I still don't think replacing `GetScaledXXX()` with `GetXXX()` is the right thing to do. This code should be valid for all ports, and it's not correct for Mac (it just so happens that we don't do any scaling for Mac neither, but this means that the correctness of this code depends on the presence of an ifdef in a very different place -- this is very fragile and confusing). I think we actually may need to use `FromDIP(GetScaledXXX())` which seems a bit silly as we basically divide by the scale factor first and then multiply by it later, but I don't see how else to handle this (which is worrisome on its own...)

pkulchenko committed:
Move Mac/GTK3 check to avoid warning on unused variable.

pkulchenko committed:
Update to fix formatting style for multi-line `if`.

pkulchenko committed:
Fix width for toolbar items with dropdown when scalefactor >1.

pkulchenko commented:
I did test the changes on macOS (with Retina 2x scaling and without on two different OS versions) and on Windows8 and Windows10 with scaling ranging from 125% to 350%.

> This code should be valid for all ports, and it's not correct for Mac (it just so happens that we don't do any scaling for Mac neither, but this means that the correctness of this code depends on the presence of an ifdef in a very different place -- this is very fragile and confusing). 

How do you suggest doing this? I suspect the original logic was not correct (as it was mixing logical and physical pixels), but not sure what you think it should be replaced with. The current code works on Mac (with the default images), as I indicated earlier.

I merged #1603 changes into this ticket. I didn't want to do this, as these changes are independent and may get delayed by the discussions on this one, but if it's easier to handle this way, then so be it. I can clean up the commits a bit after we settle on the approach, but would like to move with these changes quickly if possible.

pkulchenko commented:
Just for completeness, I also tested on GTK3 with scale factors 100-150% and didn't see any issues (I couldn't test with any larger factors).

With respect to your `FromDIP(GetScaledXXX())` idea, I don't think it will work properly, but I can give it a try. The reason for my concern is that the default images are not going to be scaled in for some scale factors (like 150%; and believe me, you don't want the scaling applied to those bitmaps with 125-175%), but FromDIP will still be applied, which may produce wrong results (which doesn't happen with direct `GetXXX` calls. I also don't expect issues with macOS, unless there are some specific scenarios where you think it may be incorrect.

pkulchenko commented:
@vadz, I tested three combinations of changes -- (1) the current code, (2) all-scaled code (I restored all `GetScaled*` calls that have been removed), and (3) all non-scaled code (I replaced all `GetScaled*` calls with `Get*` calls in tabaui.cpp) -- and don't see any difference on Windows with 125-300% scale factors and on macOS with Retina display and `GetContentScaleFactor==2`.

All the buttons and icons in notebook tabs are shown correctly. I'm not sure what specific scenarios I should be testing for that would demonstrate any difference, but this is what I've seen so far.

pkulchenko reviewed (commented):
This has been address in submitted changes and recent comments.

vadz commented:
Thanks for the update and testing, I'll try to get back to it, but I'm not sure if I can do it before 3.1.3 to be perfectly honest.

pkulchenko commented:
@vadz, I'd very much like to get these changes into the codebase before 3.1.3. If there are any specific changes you'd like me to test, I can do it later today and tomorrow.

If it makes it more comfortable for you, I can restore all `GetScaled*` calls and retest the changes on Windows and macOS. I don't see any other potential reason to delay the PR.

Also, this potential delay was one of the reasons why I wanted to get auibar fix in, as it is independent from the rest of the changes in this PR.

pkulchenko commented:
@vadz, I just realized why the tests showed no difference on macOS. It's because the changes affect scaled bitmaps and I only used regular bitmaps. Is there an easy way to emulate scaled bitmaps using regular ones? I just want to see what the effect is going to be on the drawn UI elements...

pkulchenko committed:
Update to use `GetScaled*` methods in auiTabArt calculations.

pkulchenko commented:
@vadz, I pushed one more change that restores `GetScaled*` calls and removes two `FromDIP` calls that generated excessive spacing. I think this should be closer to the original intent with much more clear results. Let me know if you'd like Windows/macOS screenshots with x2 scale.

pkulchenko commented:
@vadz, I tested on macOS with Retina display and ContentScaleFactor==2. I generated bitmaps with ScaleFactor==2 and used those as tab icons. I don't see any issues and can attach several screenshots to show how it looks at x2 factor on macOS and Windows.

I think the changes are ready to be merged.

pkulchenko commented:
@vadz, is there still an issue with the PR you'd like to address or more testing to do? I think I've addressed all your comments/concerns, but let me know if I've missed something.

vadz commented:
No, sorry, I just don't have enough time. I'll try to look at it soon but in the meanwhile it would be great if others could check it, both the code and the run-time behaviour. TIA!

MaartenBent commented:
Tab art looks better now (tested on Windows). Maybe you can do the same for dock art?


Adding/replacing  these in `wxAuiDefaultDockArt::DrawPaneButton` improves it similarly.
```c++
bmp = wxAuiScaleBitmap(bmp, window->GetContentScaleFactor());
```
```c++
dc.DrawRectangle(rect.x, rect.y, bmp.GetScaledWidth() - window->FromDIP(1), bmp.GetScaledHeight() - window->FromDIP(1));
```

pkulchenko commented:
@MaartenBent, sure thing; will do that as well shortly. Thank you for the suggestion.

@vadz, agree on the feedback wholeheartedly. I just think it would be unfortunate to not include these changes in a release that incorporates some many hidpi improvements.

pkulchenko commented:
BTW, I also confirmed the fix working as expected on GTK3 with 2x scaling.

pkulchenko committed:
Add scaling of dock art as required by scaling.

pkulchenko commented:
@MaartenBent, I added the suggested patch, but I don't see the difference between the current and patched states on Windows with 2x and 3x scaling. In both cases it looks like the dock art icons are being scaled. Can you confirm?

pkulchenko commented:
@MaartenBent, looks good; thank you for the feedback.

@vadz, I don't have any other pending changes or tests for this PR.

vadz commented:
I'm going to merge it with some very minor tweaks, thanks for the changes, but it seems that AUI is still very broken with per-monitor DPI. Starting it on 100% DPI monitor and moving it to 200% DPI completely breaks just about everything (all toolbar are corrupted, only half of the notebook tabs is shown...) but even starting it on 200% DPI shows quite a few problems, see the marked (in a highly artistic way, sorry in advance for your eyes) problems on this screenshot:

![aui_hidpi](https://user-images.githubusercontent.com/146917/67530779-7e9ec980-f6c0-11e9-8b18-a93b84300a08.png)

So I don't think AUI is ready for prime-time just yet, even with this patch. But this PR is definitely a step in the right direction, thanks again!

pkulchenko commented:
@vadz, thank you for merging!

FYI, I don't see any of those defects you are showing in the screenshot. I tested 200-300% on Win8 and my caption icons look like the ones in @MaartenBent's second screenshot (nicely positioned and proportional; although my caption text looks jagged, like it's been scaled, but is hasn't been affected by the PR).

Also issue 3 would be related to the text control, not wxAUI, right? (although, I don't see it happening either). I haven't tried dragging windows between screens with different scaling, just opened auidemo on screens with different scaling factors.

Are you testing on Win10?

vadz commented:
Yes, this is under Windows 10 1803 with 2 displays, one using 100% DPI and another 200% DPI.

MaartenBent commented:
You say per-monitor DPI (`<dpiAwareness>PerMonitorV2</dpiAwareness>`?). This PR only addresses high-DPI (`<dpiAware>true</dpiAware>` or `<dpiAwareness>system</dpiAwareness>`). I'm aware that per-monitor has more issues that need fixing.

vadz commented:
Yes, I do use `PerMonitorV2` for testing, and I understand that things don't work when moving wxAUI windows from one monitor to another. I'm surprised they don't work when starting it on a 200% DPI monitor however, especially if they do work with simple DPI awareness. I didn't realize that enabling per-monitor v2 awareness could make things worse.

BTW, I'm currently doing this
```
mt -nologo -manifest vc14x_mswud/auidemo.exe.manifest -manifest ../../../permonitorv2.manifest -outputresource:'vc14x_mswud/auidemo.exe;#1'
```
to adjust the manifest for per-monitor v2 awareness after building it using `nmake -f makefile.vc COMPILER_PREFIX=vc14x`. I hope this is correct, i.e. that the problems I'm seeing are not due to just doing something wrong in the manifest.

pkulchenko commented:
@vadz, I tested on Win10 with `PerMonitorV2` manifest, 300% scaling, *one* monitor configuration (I don't have a way to attach the second monitor) and don't see any difference between `PerMonitorV2` and `dpiAware`.

Here are my two manifests:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
  <assemblyIdentity
    version="0.0.1.1"
    name="ZeroBraneLLC.ZeroBrane.ZeroBraneStudio"
    type="win32"
  />
  <description>ZeroBrane Studio IDE</description>
  <trustInfo xmlns="urn:schemas-microsoft-com:asm.v3">
    <security>
      <requestedPrivileges>
        <requestedExecutionLevel level="asInvoker" uiAccess="false"></requestedExecutionLevel>
      </requestedPrivileges>
    </security>
  </trustInfo>
  <application xmlns="urn:schemas-microsoft-com:asm.v3">
    <windowsSettings>
      <dpiAware xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true</dpiAware>
    </windowsSettings>
  </application>
</assembly>
```

and 

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
  <assemblyIdentity
    version="0.0.1.1"
    name="ZeroBraneLLC.ZeroBrane.ZeroBraneStudio"
    type="win32"
  />
  <description>ZeroBrane Studio IDE</description>
  <trustInfo xmlns="urn:schemas-microsoft-com:asm.v3">
    <security>
      <requestedPrivileges>
        <requestedExecutionLevel level="asInvoker" uiAccess="false"></requestedExecutionLevel>
      </requestedPrivileges>
    </security>
  </trustInfo>
  <application xmlns="urn:schemas-microsoft-com:asm.v3">
    <windowsSettings>
      <dpiAwareness xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">PerMonitorV2</dpiAwareness>
    </windowsSettings>
  </application>
</assembly>
```

I'm not sure how your manifest looks like, but note that `PerMonitorV2` is mutually exclusive with `dpiaware` setting (as per the documentation: https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/mt846517(v%3Dvs.85)).

I've attached these manifests using `mt -manifest file.manifest -outputresource:auidemo.exe` command.

pkulchenko commented:
BTW, shouldn't matter, but in my case the manifest is embedded in the executable (so I can switch between different versions/configurations).
